### ISSUE TITLE
Some literal for "Null" is needed.

### ISSUE THREAD

**Issue by rhpvorderman** [openwdl/wdl#209]
I cannot find it anywhere in the spec. 
But suppose you have a task with the following inputs:
```WDL
task someTask {
  File read1
  File? read2
  String read1output
  String? read2output
  command {
    echo "some task"
  }
}
```
Now I want to run a workflow that processes a read1 and an optional read2
 Most programs will crash when a read2output is defined while no paired-end input is given. 
So in a workflow I want to do this:
```WDL
workflow someWf {
  File read1
  File? read2
  File outputDir
  call someTask {
     input:
       read1 = read1
       read2 = read2
       read1output = outputDir + "/processedRead1.fastq.gz"
       read2output = if (defined(read2)) then outputDir + "/processedRead2.fastq.gz" else NULL
  } 
}
```
But this is not possible so I do `else read2` (since read2 should evaluate to NULL if it is not present). But I don't think this is a proper way of doing this.

**Comment by cjllanwarne** [comment]
I think this is a great idea.

My only comment is that I think I'd prefer something other than `null` because of its connotations from other languages.

What I mean is, in most languages, you can assign `null` anything to anything, but in WDL you should only be able to assign this value to an optional. Maybe something like `NONE`, `EMPTY` or `UNSET` would be better?

**Comment by patmagee** [comment]
I would favor either `NONE` or `UNSET`. `EMPTY` to me implies an array or map type of object.

**Comment by pshapiro4broad** [comment]
> What I mean is, in most languages, you can assign null anything to anything, but in WDL you should only be able to assign this value to an optional. Maybe something like NONE, EMPTY or UNSET would be better?

This is not true of languages that have nullable types. In those languages (such as `C#`, which uses the same syntax to declare a nullable type) it's a compile time error to assign null (or any nullable typed expression) to a non-nullable type.

**Comment by rhpvorderman** [comment]
I would prefer `NONE` that is what is used in python and therefore  most recognizable to most scientific users, and frankly most other users as well. Also `NONE` is used in scala as well, so I presume it is a nice thing for those who write the execution engine as well.

**Comment by patmagee** [comment]
I do like `NONE`. in terms of the ease of use, I think most people would understand what `NONE` meant, without much explanation. 

I am not sure though if this needs to stand up to equality tests. ie whether we need to do a `x == NONE` or `x  == null`. IMO thats captured by `defined(x)`. `NONE` to me is a placeholder to state taht a variable does not have a value

**Comment by geoffjentry** [comment]
Something that I think R got right was separating the concept of `NULL` from `NA`, where the latter implies missing data and the former is, well, [the billion dollar mistake](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare). It's also more concise than `NONE`, and a horde of data science/comp bio types are familiar w/ R.

**Comment by pshapiro4broad** [comment]
> Something that I think R got right was separating the concept of NULL from NA, where the latter implies missing data and the former is, well, the billion dollar mistake. It's also more concise than NONE, and a horde of data science/comp bio types are familiar w/ R.

If WDL doesn't have two distinct concepts than there's nothing to separate. I've never used R so `NA` seems really odd to me, but I am probably not a typical WDL writer.

Isn't `null` used in a JSON input file to represent an input that's not provided? If there's a different term for the unset state in WDL, what if a user tries to use that term in their JSON? Using the same term in both would be less confusing, IMO.

Regarding equality testing, I think it would be better if "the unset object" behaved like any other literal in WDL and there wasn't a restriction on where it could occur.

**Comment by geoffjentry** [comment]
>  but I am probably not a typical WDL writer.

Meta: this captures one of my larger concerns about these discussions in general. Most of the folks who participate are programmers by background, and the target audience of WDL are not.

**Comment by rhpvorderman** [comment]
> Most of the folks who participate are programmers by background, and the target audience of WDL are not.

In our institute, biologists who work on computationally intensive projects have some experience scripting in python and bash. I would argue that while most of it are not programmers by background, most of the target audience at least has some programming experience. So it is not a black and white thing. In other words, putting some "programmer" ways of thinking in WDL will not scare away the target audience. As long as we do not put things in like classes, objects, recursive funtions etc.

**Comment by pshapiro4broad** [comment]
> Meta: this captures one of my larger concerns about these discussions in general. Most of the folks who participate are programmers by background, and the target audience of WDL are not.

The lack of information about the target audience isn't in itself an argument for or against any of the proposals here. If we need more information about what a typical user might expect, and this information is required to make a decision, someone will need to do the work to ensure that the terms used will be easy for users to understand.

On the other hand, when language semantics, consistency, etc, are being discussed, you'd want to use the expertise of those who have knowledge about designing languages, not typical users.

**Comment by geoffjentry** [comment]
@pshapiro4broad I'd argue you're falling into the same trap I describe, thinking of WDL as a programming language.

**Comment by rhpvorderman** [comment]
This discussion is getting of track. All I want is an option to say: if this condition, then define this optional value, else do not define it (Scala's `None`). It does not necessarily have to be a literal for null. This feature is useful whether you are programmer or not. It happens a lot with programs that accept both paired-end and single-end reads.

**Comment by patmagee** [comment]
Bringing this back to relevance, I wanted to reiterate that I support the use of `None`. It seems like this is one of the least controversial ways of defining an empty input.

**Comment by cjllanwarne** [comment]
This looks like another situation where some "None" value would have been useful in a real world situation: https://gatkforums.broadinstitute.org/firecloud/discussion/10740/error-the-local-copy-message-must-have-path-set

**Comment by rhpvorderman** [comment]
Another situation:
A file is created dependent on some command flag.
```WDL
output {
  File? conditionalOutput = if (commandFlag) then "path/to/conditional.file" else ???
}
```
What should I put at `???`?
As a workaround I can create a `File? noneFile` in the input section. But that is not ideal as the user can overwrite this.
Any other existing workarounds for this in cromwell @cjllanwarne @geoffjentry ?

**Comment by yfarjoun** [comment]
what happens if you drop the else clause altogether?
```wdl
output {
  File? conditionalOutput = if (commandFlag) then "path/to/conditional.file"
}
```

**Comment by rhpvorderman** [comment]
@yfarjoun  then the WDL parser complains that it needs an else clause.

This is another debate: do we need an else clause if it will say: `else None`. Do we even need `None` when we can drop the else clause.

My stance on this is: Explicit is better than Implicit. Especially if WDL aims to be human-readable. As you simply can not read things that are implicit. So I would favour having `else None` over not having it at all. Optionals are something that takes some time to get yourself adjusted to if you have not seen them before. If WDL becomes riddled with implicits concerning optionals I fear that it will become much less accessible.

**Comment by yfarjoun** [comment]
there already are implicits, for example 

```wdl
workflow someWf{
boolean flag

if (flag) {
   call someTask
}

```

Now `someTask` has optional outputs, implicitly, even if it's outputs were not optional to begin with.

**Comment by rspreafico-vir** [comment]
Upvoting this feature. Having optional inputs but not having an explicit Null/None severely limits what can be done and forces using weird/ugly constructs to get around the limitation.

**Comment by mlin** [comment]
@rspreafico-vir FYI the spec change is in as of recently (i.e. not yet available in released software versions, but coming) https://github.com/openwdl/wdl/pull/263

**Comment by geoffjentry** [comment]
This was implemented in #263
